# Beer Analysis

In [1]:
import pandas as pd 
import numpy as nm
import matplotlib.pyplot as plt

In [2]:
beer_data = pd.read_csv("beer_reviews.csv")

## Cleaning Data

In [3]:
beer_data.shape

(1048575, 13)

In [4]:
deDuplicates = beer_data.drop_duplicates()

In [5]:
deDuplicates.shape

(1048575, 13)

In [6]:
beer_data = beer_data.dropna()

In [7]:
beer_data.shape

(1005519, 13)

## Question 1:

In [8]:
max(beer_data['beer_abv'])

57.7

In [9]:
beer_data[beer_data['beer_abv'] == 57.7]

brewery_id  brewery_name  review_time  review_overall  review_aroma  \
12919        6513  Schorschbräu   1316780901             4.0           4.0   

       review_appearance review_profilename beer_style  review_palate  \
12919                4.0        kappldav123    Eisbock            4.0   

       review_taste                      beer_name  beer_abv  beer_beerid  
12919           3.5  Schorschbräu Schorschbock 57%      57.7        73368

## Comments:

The Strongest beer will have the highest ABV percentage. According to that, The highest ABV percentage is 57.7 which is present in the brewery "Schorschbräu" 

## Question 2:

In [10]:
max(beer_data['review_overall'])

5.0

In [11]:
overall = beer_data['review_overall'] == 5.0

In [12]:
aroma = beer_data['review_aroma'] == 5.0

In [13]:
appearance = beer_data['review_appearance'] == 5.0

In [14]:
palate = beer_data['review_palate'] == 5.0 

In [15]:
taste = beer_data['review_taste'] == 5.0

In [16]:
best_rated = beer_data[overall & aroma & appearance & palate & taste]

In [17]:
best_rated_sorted = (best_rated.groupby('beer_beerid')['beer_beerid'].count())

In [20]:
best_rated_sorted = pd.DataFrame({'beerid':best_rated_sorted.index, 'count':best_rated_sorted.values})

In [21]:
best_rated_sorted.sort_values('count',ascending = False)

beerid  count
518    19960    110
71       645     84
441    11757     72
494    17112     72
128     1160     68
80       695     55
90       731     55
877    47658     50
473    15881     44
438    10672     44
121     1093     42
19       129     38
249     2512     35
619    29619     34
20       131     33
3         34     33
154     1444     32
309     4083     30
432    10325     30
603    28203     29
206     1904     29
1019   57747     28
119     1056     28
1009   56973     27
53       571     26
257     2671     24
647    33127     23
28       224     23
862    47022     22
358     6368     21
...      ...    ...
559    23474      1
560    23495      1
527    20564      1
524    20517      1
484    16491      1
522    20310      1
485    16506      1
486    16558      1
489    16971      1
490    17015      1
491    17032      1
493    17062      1
495    17141      1
496    17206      1
497    17226      1
498    17477      1
501    17647      1
504    18471      1
506    18698      1
509    18876      1
511    19324      1
512    19363      1
513    19398      1
514    19488      1
515    19680      1
516    19796      1
517    19944      1
519    20014      1
521    20298      1
1179   77087      1

[1180 rows x 2 columns]

In [22]:
 best_rated[best_rated['beer_beerid'] == 11757]

brewery_id              brewery_name  review_time  review_overall  \
21396        1199  Founders Brewing Company   1274055673             5.0   
21497        1199  Founders Brewing Company   1266035267             5.0   
21551        1199  Founders Brewing Company   1262480849             5.0   
21556        1199  Founders Brewing Company   1262302841             5.0   
21705        1199  Founders Brewing Company   1322237720             5.0   
21773        1199  Founders Brewing Company   1253063048             5.0   
21808        1199  Founders Brewing Company   1322187317             5.0   
21851        1199  Founders Brewing Company   1248491042             5.0   
21912        1199  Founders Brewing Company   1241833826             5.0   
21921        1199  Founders Brewing Company   1321805387             5.0   
21932        1199  Founders Brewing Company   1241309486             5.0   
21979        1199  Founders Brewing Company   1238592642             5.0   
22034        1199  Founders Brewing Company   1321332178             5.0   
22115        1199  Founders Brewing Company   1232162292             5.0   
22136        1199  Founders Brewing Company   1321224870             5.0   
22220        1199  Founders Brewing Company   1228688329             5.0   
22237        1199  Founders Brewing Company   1319502328             5.0   
22240        1199  Founders Brewing Company   1227802195             5.0   
22260        1199  Founders Brewing Company   1227433878             5.0   
22304        1199  Founders Brewing Company   1225847899             5.0   
22356        1199  Founders Brewing Company   1320810742             5.0   
22376        1199  Founders Brewing Company   1223514729             5.0   
22396        1199  Founders Brewing Company   1223082826             5.0   
22402        1199  Founders Brewing Company   1222567348             5.0   
22414        1199  Founders Brewing Company   1221964620             5.0   
22459        1199  Founders Brewing Company   1213303453             5.0   
22561        1199  Founders Brewing Company   1199843255             5.0   
22572        1199  Founders Brewing Company   1199221866             5.0   
22646        1199  Founders Brewing Company   1193368601             5.0   
22739        1199  Founders Brewing Company   1169430192             5.0   
...           ...                       ...          ...             ...   
23073        1199  Founders Brewing Company   1318654845             5.0   
23090        1199  Founders Brewing Company   1318222551             5.0   
23162        1199  Founders Brewing Company   1316626780             5.0   
23166        1199  Founders Brewing Company   1316399763             5.0   
23173        1199  Founders Brewing Company   1316289259             5.0   
23174        1199  Founders Brewing Company   1316221799             5.0   
23176        1199  Founders Brewing Company   1316215428             5.0   
23191        1199  Founders Brewing Company   1315890993             5.0   
23199        1199  Founders Brewing Company   1315520394             5.0   
23236        1199  Founders Brewing Company   1314391659             5.0   
23238        1199  Founders Brewing Company   1314321377             5.0   
23255        1199  Founders Brewing Company   1313111154             5.0   
23269        1199  Founders Brewing Company   1312328160             5.0   
23285        1199  Founders Brewing Company   1309223278             5.0   
23288        1199  Founders Brewing Company   1308930319             5.0   
23309        1199  Founders Brewing Company   1305679635             5.0   
23317        1199  Founders Brewing Company   1305072552             5.0   
23329        1199  Founders Brewing Company   1325477496             5.0   
23330        1199  Founders Brewing Company   1325446649             5.0   
23368        1199  Founders Brewing Company   1301371986             5.0   
23464        1199  Founders Brewing Company   129565422

In [23]:
 best_rated[best_rated['beer_beerid'] == 17112]

brewery_id          brewery_name  review_time  review_overall  \
930964         287  Bell's Brewery, Inc.   1264906401             5.0   
931105         287  Bell's Brewery, Inc.   1263609584             5.0   
931129         287  Bell's Brewery, Inc.   1263434423             5.0   
931150         287  Bell's Brewery, Inc.   1263166437             5.0   
931165         287  Bell's Brewery, Inc.   1259534029             5.0   
931202         287  Bell's Brewery, Inc.   1248736433             5.0   
931208         287  Bell's Brewery, Inc.   1248294172             5.0   
931281         287  Bell's Brewery, Inc.   1302577635             5.0   
931444         287  Bell's Brewery, Inc.   1235787916             5.0   
931482         287  Bell's Brewery, Inc.   1235070725             5.0   
931487         287  Bell's Brewery, Inc.   1234997663             5.0   
931494         287  Bell's Brewery, Inc.   1234896720             5.0   
931511         287  Bell's Brewery, Inc.   1301440626             5.0   
931529         287  Bell's Brewery, Inc.   1234419511             5.0   
931544         287  Bell's Brewery, Inc.   1234208576             5.0   
931546         287  Bell's Brewery, Inc.   1234160274             5.0   
931623         287  Bell's Brewery, Inc.   1300917463             5.0   
931662         287  Bell's Brewery, Inc.   1232849071             5.0   
931689         287  Bell's Brewery, Inc.   1232608751             5.0   
931704         287  Bell's Brewery, Inc.   1232416926             5.0   
931720         287  Bell's Brewery, Inc.   1232133944             5.0   
931721         287  Bell's Brewery, Inc.   1232084081             5.0   
931736         287  Bell's Brewery, Inc.   1300669760             5.0   
931751         287  Bell's Brewery, Inc.   1227221316             5.0   
931780         287  Bell's Brewery, Inc.   1219895217             5.0   
931855         287  Bell's Brewery, Inc.   1210586736             5.0   
931870         287  Bell's Brewery, Inc.   1209260993             5.0   
931956         287  Bell's Brewery, Inc.   1300390010             5.0   
932050         287  Bell's Brewery, Inc.   1202843507             5.0   
932065         287  Bell's Brewery, Inc.   1202690670             5.0   
...            ...                   ...          ...             ...   
932554         287  Bell's Brewery, Inc.   1163607214             5.0   
932570         287  Bell's Brewery, Inc.   1163222335             5.0   
932587         287  Bell's Brewery, Inc.   1162663891             5.0   
932599         287  Bell's Brewery, Inc.   1298768056             5.0   
932617         287  Bell's Brewery, Inc.   1298582435             5.0   
932705         287  Bell's Brewery, Inc.   1297203676             5.0   
932730         287  Bell's Brewery, Inc.   1296955585             5.0   
932735         287  Bell's Brewery, Inc.   1296928471             5.0   
932757         287  Bell's Brewery, Inc.   1296650357             5.0   
932770         287  Bell's Brewery, Inc.   1314319294             5.0   
932771         287  Bell's Brewery, Inc.   1313074760             5.0   
932775         287  Bell's Brewery, Inc.   1312423007             5.0   
932792         287  Bell's Brewery, Inc.   1296434727             5.0   
932813         287  Bell's Brewery, Inc.   1295633670             5.0   
932856         287  Bell's Brewery, Inc.   1296077940             5.0   
932944         287  Bell's Brewery, Inc.   1295674654             5.0   
932957         287  Bell's Brewery, Inc.   1295652496             5.0   
933070         287  Bell's Brewery, Inc.   1295066015             5.0   
933078         287  Bell's Brewery, Inc.   1295055938             5.0   
933106         287  Bell's Brewery, Inc.   1308595695             5.0   
933119         287  Bell's Brewery, Inc.   1294808960             5.0   
933154         287  Bell's Brewery, Inc.   1283023546             5.0   
933168         287  Bell's Brewery, Inc.   1278474658             

## Comments:

Taken a frequency measure of beers with all 5.0 ratings (overall,appearence,aroma,palate,taste) and selected the top 3 <br>
  beer_id count <br>
1.19960   112 <br>
2.645     84 <br>
3.11757   72 <br>
3.17112   72 <br>
To break the tie between the beers in the 3rd position, considered ABV percentage. <br>
So, 17112 has the highest ABV <br>
Finally, Suggesting the top 3 beers, <br>
1. 19960 <br>
2. 645 <br>
3. 17112

## Question 3:

In [24]:
beer_data["review_overall"].corr(beer_data['review_aroma'])

0.5993713657013837

In [25]:
beer_data["review_overall"].corr(beer_data['review_appearance'])

0.48102808315629453

In [26]:
beer_data["review_overall"].corr(beer_data['review_palate'])

0.6887532862291011

In [27]:
beer_data["review_overall"].corr(beer_data['review_taste'])

0.7811639448182778

## Comments:

To find the most important factor in determining the overall quality of beer, correlation of overall rating with rest fours ratings is taken. <br>
From the results found, We can say that taste, palate, aroma, appearance determines the overall rating in order. <br>
Taste is the most important factor.

## Question 4:

In [28]:
aroma_appearance = beer_data[aroma & appearance]

In [29]:
aroma_appearance

brewery_id                                       brewery_name  \
287            1075                            Caldera Brewing Company   
443            1075                            Caldera Brewing Company   
580            1075                            Caldera Brewing Company   
906            1075                            Caldera Brewing Company   
1697           1454                              Broad Ripple Brew Pub   
1729           1454                              Broad Ripple Brew Pub   
1768           1454                              Broad Ripple Brew Pub   
2008            577                               Isle Of Skye Brewery   
2130            577                               Isle Of Skye Brewery   
2399          16859                            Great South Bay Brewery   
2878          14879                               Hoppin' Frog Brewery   
2936          14879                               Hoppin' Frog Brewery   
3146          14879                               Hoppin' Frog Brewery   
3250          14879                               Hoppin' Frog Brewery   
3281          14879                               Hoppin' Frog Brewery   
3373          14879                               Hoppin' Frog Brewery   
3389          14879                               Hoppin' Frog Brewery   
3665          14879                               Hoppin' Frog Brewery   
3678          14879                               Hoppin' Frog Brewery   
3685          14879                               Hoppin' Frog Brewery   
3705          14879                               Hoppin' Frog Brewery   
3713          14879                               Hoppin' Frog Brewery   
3717          14879                               Hoppin' Frog Brewery   
3731          14879                               Hoppin' Frog Brewery   
3756          14879                               Hoppin' Frog Brewery   
4411          14879                               Hoppin' Frog Brewery   
4562          14879                               Hoppin' Frog Brewery   
4691          14879                               Hoppin' Frog Brewery   
4722          14879                               Hoppin' Frog Brewery   
4724          14879                               Hoppin' Frog Brewery   
...             ...                                                ...   
1041620         173  Hair of the Dog Brewing Company / Brewery and ...   
1041704         173  Hair of the Dog Brewing Company / Brewery and ...   
1041718         173  Hair of the Dog Brewing Company / Brewery and ...   
1041876         173  Hair of the Dog Brewing Company / Brewery and ...   
1041891         173  Hair of the Dog Brewing Company / Brewery and ...   
1042110         173  Hair of the Dog Brewing Company / Brewery and ...   
1042155         173  Hair of the Dog Brewing Company / Brewery and ...   
1042174         173  Hair of the Dog Brewing Company / Brewery and ...   
1042194         173  Hair of the Dog Brewing Company / Brewery and ...   
1042245         173  Hair of the Dog Brewing Company / Brewery and ...   
1042286         173  Hair of the Dog Brewing Company / Brewery and ...   
1042308         173  Hair of the Dog Brewing Company / Brewery and ...   
1042329         173  Hair of the Dog Brewing Company / Brewery and ...   
1042334         173  Hair of the Dog Brewing Company / Brewery and ...   
1042408         173  Hair of the Dog Brewing Company / Brewery and ...   
1042409         173  Hair of the Dog Brewing Company / Brewery and ...   
1042595         173  Hair of the Dog Brewing Company / Brewery and ...   
1042744         173  Hair of the Dog Brewing Company / Brewery and ...   
1043928         726                   Dortmunder Actien Brauerei / DAB   
1044208         726                   Dortmunder Actien Brauerei / DAB   
1044404         257                              Steel Brewing Company   
1044431         257                              Steel Brewing Company   
1045348      

In [32]:
aroma_appearance_sorted = (aroma_appearance.groupby('beer_style')['beer_style'].count())

In [33]:
aroma_appearance_sorted = pd.DataFrame({'beer_style':aroma_appearance_sorted.index, 'count':aroma_appearance_sorted.values})

In [34]:
aroma_appearance_sorted.sort_values('count',ascending = False)

beer_style  count
10     American Double / Imperial Stout   1368
84               Russian Imperial Stout    850
11                         American IPA    840
8        American Double / Imperial IPA    811
18                  American Strong Ale    363
24              Belgian Strong Dark Ale    293
82                     Quadrupel (Quad)    251
25              Belgian Strong Pale Ale    230
92                               Tripel    228
61                           Hefeweizen    214
4                   American Barleywine    203
16                      American Porter    180
36                           Doppelbock    177
2              American Amber / Red Ale    175
17                       American Stout    162
86               Saison / Farmhouse Ale    161
70                       Lambic - Fruit    158
23                     Belgian Pale Ale    143
38                               Dubbel    140
79                        Oatmeal Stout    130
13              American Pale Ale (APA)    121
41                   English Barleywine    119
58               Fruit / Vegetable Beer    105
19                    American Wild Ale     98
94                           Weizenbock     94
85                             Rye Beer     85
60                               Gueuze     83
5                    American Black Ale     72
20                        Baltic Porter     72
80                              Old Ale     66
..                                  ...    ...
69                               Kölsch     12
76                  Munich Dunkel Lager     12
71                   Lambic - Unblended     12
54  Extra Special / Strong Bitter (ESB)     11
72                          Light Lager     11
42                       English Bitter     10
64                        Irish Red Ale     10
14                  American Pale Lager      9
3            American Amber / Red Lager      9
50                   English Strong Ale      8
12                 American Malt Liquor      7
52                      Euro Pale Lager      5
6                   American Blonde Ale      4
26                   Berliner Weissbier      4
91                          Smoked Beer      4
49                        English Stout      4
31                              Braggot      4
67                        Kristalweizen      4
51                      Euro Dark Lager      3
32       California Common / Steam Beer      2
68                                Kvass      2
66           Keller Bier / Zwickel Bier      2
65                  Japanese Rice Lager      2
44                English Dark Mild Ale      2
53                    Euro Strong Lager      2
29                          Black & Tan      1
33                           Chile Beer      1
73                     Low Alcohol Beer      1
90  Scottish Gruit / Ancient Herbed Ale      1
47                English Pale Mild Ale      1

[98 rows x 2 columns]

## Comments:

Taken a frequency measure of beers with all 5.0 ratings (appearance, aroma) and selected the best beer style which is American Double /aImperial Stout.